app/database/operator/server_data_operator/handlers/group.ts

```diff
@@ -41,6 +45,51 @@ const GroupHandler = (superclass: any) => class extends superclass implements Gr
             prepareRecordsOnly,
         });
     };
+
+    /**
+     * handleGroupMembershipsForGroup: Handler responsible for the Create/Update operations occurring on the GroupMembership table from the 'Server' schema
+     * @param {string} userId
+     * @param {HandleGroupMembershipForMemberArgs} groupMembershipsArgs
+     * @param {GroupMembership[]} groupMembershipsArgs.groupMemberships
+     * @param {boolean} groupMembershipsArgs.prepareRecordsOnly
+     * @throws DataOperatorException
+     * @returns {Promise<GroupMembershipModel[]>}
+     */
+    handleGroupMembershipsForMember = async ({userId, groups, prepareRecordsOnly = true}: HandleGroupMembershipForMemberArgs): Promise<GroupMembershipModel[]> => {
+        // Get existing group memberships
+        const existingGroupMemberships = await queryGroupMembershipForMember(this.database, userId).fetch();
+
+        // Prune memberships that are no longer required
+        const membershipsToRemove = existingGroupMemberships.
+            filter((gm) => !groups?.some((g) => gm.groupId === g.id)).
```

Comment:
You have in this file several nested loops (`filter(() => some())` => O(n2)). 0/5 on whether we want to try to keep it on O(n) (by using Set, objects, or whatever other tool).

Reply:
I think I can cut down on the loops a bit as it is 2 * O(3) right now (filter against array then a map, for two sets of arrays); but I'm not sure if there's an elegant way of cutting on down on the iterations for the last scenario -

- If no group-memberships from the API, then remove existing ones; single iteration for O(1)
- If no existing group-memberships, add all the ones from the API; single iteration for O(1)
- If existing group-memberships, and groups: remove any existing group-memberships that are not in the groups (O(n2)) via `.prepareDestroy()`, remove intersection of groups + group-memberships (O(n2)) as we don't want unnecessary updates, transform remaining groups to group-memberships (O(n1)) and run handler, batch existing group-memberships (should only be deletes left) + new inserts (should only be create's left)

Reply:
Oh and the `Group[]` here represents the most up to date memberships; as that array is all the groups that the user is a member of.

Reply:
Maybe something like, it cuts it down to O(2) + O(1):

```
        const existingGroupsToRemove = [];
        for (const gm of existingGroupMemberships) {
            const index = groups.findIndex((g) => g.id === gm.groupId);

            if (index > -1) {
                groups.splice(index, 1);
            } else {
                existingGroupsToRemove.push(gm.prepareDestroyPermanently());
            }
        }

        const groupsToInsert = groups.forEach((g) => ({id: `${g.id}-${userId}`, user_id: userId, group_id: g.id}));
        
        ... run handler, then batch insert records
        ```

Reply:
In that option you still have a for->find, therefore you are still with O(n2).
You can create a map of groups (`{[x: string]: Group | undefined}`). Accessing a map, in theory is O(1), instead of O(n) like the find, so there is no issue adding searching in the map (i.e. `myGroupMap[gm.groupId] !== undefined`).

It is a bit cumbersome, because you have to create this map, and probably transform it back to a list (hopefully `Object.values` works good enough) but that way you make sure you reduce the complexity.

Reply:
Yup yup! Updated, so now it's just 4x O(n1) instead from what I can tell.